### PR TITLE
Migrate memory_snapshot_models.dart to null safety.

### PR DESCRIPTION
**Pull request by polina-c**

### CODE REVIEW

File: packages/devtools_app/lib/src/screens/memory/memory_snapshot_models.dart [hunk 1]
```diff
@@ -943,9 +945,10 @@ FieldReference listToFieldEntries(
               );
 
               FieldReference valueEntry;
-              final HeapGraphClassLive valueClass = valueElement.theClass;
-              predefined = predefinedClasses[valueClass.fullQualifiedName];
-              if (predefined != null && predefined.isScalar) {
+              final HeapGraphClassLive valueClass =
+                  valueElement.theClass as HeapGraphClassLive;
+              predefined = predefinedClasses[valueClass.fullQualifiedName]!;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/memory_snapshot_models.dart:950]: maybe we should remove the null assert and go back to checking for null in the if below. This seems safer

File: packages/devtools_app/lib/src/screens/memory/memory_controller.dart [hunk 2]
```diff
@@ -1113,15 +1113,18 @@ class MemoryController extends DisposableController
       ExternalReference? externalReference;
 
       if (externalReferences.children.isNotEmpty) {
-        externalReference = externalReferences.children.singleWhere(
-          (knownClass) => knownClass.name == classLive?.name,
+        externalReference =
+            // We need cast to return null.
+            // ignore: unnecessary_cast
+            (externalReferences.children as List<Reference?>).singleWhere(
```
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/memory_controller.dart:1119]: it was side effect of casting

[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/memory_controller.dart:1119]: done

File: packages/devtools_app/lib/src/screens/memory/memory_snapshot_models.dart [hunk 3]
```diff
@@ -91,7 +89,7 @@ class Reference extends TreeNode<Reference> {
         );
 
   Reference.allocationMonitor(MemoryController controller, String name,
-      {Function onExpand, Function onLeaf})
+      {Function? onExpand, Function? onLeaf})
```
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/memory_snapshot_models.dart:92]: done

File: packages/devtools_app/lib/src/screens/memory/memory_snapshot_models.dart [hunk 4]
```diff
@@ -488,18 +484,18 @@ class ClassReference extends Reference {
           },
         );
 
-  List<HeapGraphElementLive> get instances =>
-      actualClass.getInstances(controller.heapGraph);
+  List<HeapGraphElementLive?> get instances =>
```
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/memory_snapshot_models.dart:487]: done

File: packages/devtools_app/lib/src/screens/memory/memory_snapshot_models.dart [hunk 5]
```diff
@@ -488,18 +484,18 @@ class ClassReference extends Reference {
...
 class ObjectReference extends Reference {
   ObjectReference(
-    MemoryController controller,
+    MemoryController? controller,
     int index,
     this.instance,
   ) : super.object(controller, 'Instance $index');
 
-  final HeapGraphElementLive instance;
+  final HeapGraphElementLive? instance;
```
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/memory_snapshot_models.dart:498]: done

File: packages/devtools_app/lib/src/screens/memory/memory_snapshot_models.dart [hunk 6]
```diff
@@ -662,31 +658,36 @@ class ObjectFieldReference extends FieldReference {
...
       final computedFields = <FieldReference>[];
       for (final ref in objFields) {
         // Check if aready computed/expanded then skip.
-        final exist = objectFieldReference.children.singleWhere(
-          (element) => element.name == ref.name,
+        final FieldReference? exist =
+            // We need this cast to return null.
+            // ignore: unnecessary_cast
+            (objectFieldReference.children as List<FieldReference?>)
+                .singleWhere(
```
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/memory_snapshot_models.dart:672]: done

File: packages/devtools_app/lib/src/screens/memory/memory_snapshot_models.dart [hunk 7]
```diff
@@ -662,31 +658,36 @@ class ObjectFieldReference extends FieldReference {
...
+          (element) => element!.name == ref.name,
           orElse: () => null,
         );
         if (exist == null) {
           // Doesn't exist so add field.
           if (ref is ObjectFieldReference && !ref.isNull) {
             computedFields.add(ref);
-          } else if (ref is FieldReference) {
+          } else {
```
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/memory_snapshot_models.dart:680]: The collection that we loop (objFields) is List<FieldReference>. So, ref is always FieldReference.

File: packages/devtools_app/lib/src/screens/memory/memory_snapshot_models.dart (hunk 1 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/memory_snapshot_models.dart:950]: done

**Review by kenzieschmoll — APPROVED** [review]

**Comment by polina-c** [comment]
tested

**Pull request merged**